verifier/verifier.go

```diff
@@ -0,0 +1,142 @@
+	// validation not involving a name mismatch, e.g. if a chain could not be
+	// built.
+	ValidationError error
+
+	// NameError will be non-nil when there was a mismatch between the name on the
+	// certificate and the name being verified against.
+	NameError error
+
+	// Chains is a list of validated certificate chains, starting at the
+	// certificate being verified, and ending at a certificate in the root store.
+	Chains [][]*x509.Certificate
+}
+
+// MatchesDomain returns true if NameError == nil and Name != "".
+func (res *VerificationResult) MatchesDomain() bool {
+	return res.NameError == nil && res.Name != ""
+}
+
+// VerifyProcedure is an interface to implement additional browser specific logic at
+// the start and end of verification.
+type VerifyProcedure interface {
+	PreValidate(c *x509.Certificate) error
+	PostValidate(c *x509.Certificate, chains [][]*x509.Certificate) error
+}
+
+// VerificationOptions contains settings for Browser.Verify().
+// VerificationOptions should be safely copyable.
+type VerificationOptions struct {
+	VerifyTime time.Time
+	Name       string
+}
+
+func (opt *VerificationOptions) clean() {
+	if opt.VerifyTime.IsZero() {
+		opt.VerifyTime = time.Now()
+	}
+}
+
+// A Verifier represents a context for verifying certificates.
+type Verifier struct {
+	Roots           *x509.CertPool
+	Intermediates   *x509.CertPool
+	Blacklist       *x509.CertPool
+	Whitelist       *x509.CertPool
+	VerifyProcedure VerifyProcedure
+}
+
+func (v *Verifier) convertOptions(opt *VerificationOptions) (out x509.VerifyOptions) {
+	opt.clean()
+	out.CurrentTime = opt.VerifyTime
+	out.Roots = v.Roots
+	out.Intermediates = v.Intermediates
+	out.DNSName = opt.Name
+	return
+}
+
+// Verify checks if c chains back to a certificate in Roots, possibly via a
+// certificate in Intermediates, and returns all such chains. It additional
+// checks if the Name in the VerificationOptions matches the name on the
+// certificate. Finally, it checks to see if c is blacklisted or whitelisted by
+// the Browser.
+func (b *Browser) Verify(c *x509.Certificate, opts VerificationOptions) (res *VerificationResult) {
+	res = new(VerificationResult)
+	res.Name = opts.Name
+
+	// Always run prevalidate.
+	if err := v.VerifyProcedure.PreValidate(c); err != nil {
+		res.ValidationError = err
+	}
+
+	// XXX: x509 should expose the validation we want with out this struct
+	// designed for JSON output.
+	if res.ValidationError == nil {
+		xopts := v.convertOptions(&opts)
+		chains, validation, _ := c.ValidateWithStupidDetail(xopts)
+		res.Chains = chains
+		res.ValidationError = validation.ValidationError
+		res.NameError = validation.NameError
+		res.Chains = chains
+	}
+
+	// Run PostValidate if there was no exisiting validation error.
+	if res.ValidationError == nil {
+		if err := v.VerifyProcedure.PostValidate(c, res.Chains); err != nil {
+			res.ValidationError = err
+		}
+	}
+
+	// Check the whitelist
+	if v.Whitelist.Contains(c) {
```

Comment:
Again, we have _no_ idea if this is the right way of doing this, and right now isn't the time to be working this out.